### DISCUSSION TITLE
Can't load EFI file in iPXE that works fine in the BIOS EFI Shell

### DISCUSSION THREAD

**Discussion by bshogeman** [ipxe/ipxe#550] [Q&A] [answered]
I am trying to use iDmiEditEfi64.efi which is part of the [Intel Firmware Integrator Tools iDmi](https://www.intel.com/content/www/us/en/download/19504/intel-aptio-v-uefi-firmware-integrator-tools.html) but can't get it to work within iPXE. Other EFI files load fine. But specific this one doesn't. It is 64bit just as my system and it works if I use the Intergraded EFI shell from the BIOS. But it doesn't run from the iPXE shell or an menu. It gives first OK for loading the file followed by the error 0x7F0482FF  Is there an option in the iPXE config that I must enable before loading this kind of efi files? Chainloading or imgload doesn't make a difference.

**Comment by NiKiZe** [comment]
iPXE calls the EFI firmware which runs the binary.
One of the differences between efi shell and iPXE is that with the shell you probably get a disk as "path" while with iPXE it is always from memory.

Please post the full error message, including the iPXE.org URL that is in the message.

One workaround, if the path is important, is to (ab)use wimboot and its ability to give a virtual FS.

**Reply by NiKiZe** [reply]
The wimboot hack is essentially:
```bash
# clear any images
imgfree
# download iDmiEditEfi64.efi and keep it in memory as bootx64.efi
imgfetch --name bootx64.efi http://server/iDmiEditEfi64.efi
# use wimboot to boot from a virtual FS, gui to not destroy VESA tables
chain http://server/wimboot
```

**Reply by NiKiZe** [reply]
A workaround using wimboot + shell.efi + startup.nsh is available using chain http://b800.org/idmi/boot.ipxe

EDIT: current version of above script first uses variant that does not use wimboot as described in https://github.com/ipxe/ipxe/discussions/550#discussioncomment-1961943

Output:
```
iPXE 1.21.1+ (g91c77) -- Open Source Network Boot Firmware -- http://ipxe.org
Features: DNS HTTP iSCSI TFTP SRP AoE EFI Menu

iPXE> dhcp
Configuring (net0 52:54:00:12:34:56)...... ok
iPXE> chain http://b800.org/idmi/boot.ipxe
http://b800.org/idmi/boot.ipxe... ok
iDmiEditEfi64.efi... ok
startup.nsh... ok
http://boot.ipxe.org/Shell.efi... ok
wimboot... ok


wimboot v2.7.3 -- Windows Imaging Format bootloader -- https://ipxe.org/wimboot

Command line: "wimboot"
Using iDmiEditEfi64.efi via 0x7e2fa6d8 len 0x537f0
Using startup.nsh via 0x7e2fa798 len 0x1a
Using bootx64.efi via 0x7e2fa858 len 0xf1d80
...found bootmgfw.efi file bootx64.efi
Using wimboot via 0x7e2fa918 len 0xe0b0
Read bootx64.efi
Loaded bootx64.efi


UEFI Interactive Shell v2.2
EDK II
UEFI v2.70 (EDK II, 0x00010000)
Mapping table
      FS0: Alias(s):F1:;BLK1:
          PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)
     BLK0: Alias(s):
          PciRoot(0x0)/Pci(0x1F,0x2)/Sata(0x2,0xFFFF,0x0)
     BLK2: Alias(s):
          VenHw(1322D197-15DC-4A45-A6A4-FA57054EA614)/Ata(0x0)
     BLK3: Alias(s):
          VenHw(1322D197-15DC-4A45-A6A4-FA57054EA614)/Ata(0x0)/HD(1,MBR,0xC0FFEEEE,0x80,0xFFF7FFC0)

Press ESC in 1 seconds to skip startup.nsh or any other key to continue.
Shell> iDmiEditEfi64.efi /ca
This tool is not supported on this system.
49 - Error: A platform condition has prevented executing.
Shell> exit
bootx64.efi returned
Could not boot: Error 0x7f048281 (http://ipxe.org/7f048281)
Could not boot: Error 0x7f048281 (http://ipxe.org/7f048281)
iPXE>
```

boot.ipxe:
```bash
#!ipxe

# clear any images
imgfree

imgfetch iDmiEditEfi64.efi
# download startup.nsh for autostart in efi shell
# this can be a serverside dynamic script that takes options such as: /cs "${serial}" /ca "${asset}"
# and then generates the valid startup.nsh contents
# -n startup.nsh makes sure the local name in the fs is correct, while the server url can be anything
imgfetch -n startup.nsh startup.nsh

# download Shell.efi and keep it in memory as bootx64.efi
imgfetch -n bootx64.efi http://boot.ipxe.org/Shell.efi

# use wimboot to boot from a virtual FS, gui to not destroy VESA tables
boot wimboot gui
```

**Comment by bshogeman** [comment]
First the screenshot from the Intergrated EFI Shell that can be started from the BIOS in uEFI mode.
![IMG_Working_EFI-Shell](https://user-images.githubusercontent.com/4386297/149159714-92d06555-e107-43a9-ba02-cb8daf44215f.jpg)
And now load iPXE first and than chain the EFI file. This gives an error.
![IMG_iPXE_shell-with-error](https://user-images.githubusercontent.com/4386297/149159726-80905eab-bac8-4671-9caa-6bccb7276647.jpg)

The file does nothing more then writing an asset to the smbios

**Reply by NiKiZe** [reply]
Error: Could not boot: Error 0x7f0482ff (http://ipxe.org/7f0482ff)

**Reply by NiKiZe** [reply]
That links to https://ipxe.org/err/7f048200
This means that the binary starts fine, but returns error code 255.

**Reply by bshogeman** [reply]
Then alone what can I do about it and make it run within iPXE the same as in the EFI Shell.
Is it an option that is not configured when building the ipxe binary or must I load the efi file different within ipxe.
Errorcode 255 is even with that link to generic to find an solution.

**Reply by NiKiZe** [reply]
Some things to test, is iDmiEditEfi64.efi the only file on wherever medium you are using? (Looks like it but want to verify)
Also can you rename it bootx64.efi and it still works? (If yes then I have the wimboot hack for you coming up)

**Reply by bshogeman** [reply]
No iDmiEditEfi64.efi is not the only file on the medium. With iPXE I download the file with help of the http protocol of the server. Then it could passthrough the specified serial number and asset. iPXE first prompt to give them in and pass them trough to this EFI as parameter.
The Integrated EFI Shell doesn't support http transfer and can only open local files. Therefore for testing the file is copied to the efi boot partition and started from there. Then it works.

Wimboot.. It is of course possible to place the iDmiEditEfi64.efi in an WIM image file and then boot it with wimboot. But how do I pass parameters trough like the new asset?

#!ipxe
echo -n Nieuw serial: && read serial
echo -n Nieuw asset: && read asset
chain http://bootserver.domain.local/boot/asset/iDmiEditEfi64.efi /cs "${serial}" /ca "${asset}"

**Reply by NiKiZe** [reply]
So not sure what the answer is here? Does iDmiEditEfi64.efi work if it is the only file on the disk when run from EFI shell?
You need to get a baseline for what works and not to try and isolate the issue.

I never said anything about putting it in a .wim file - don't

Again let's start with the basic test of getting iDmiEditEfi64.efi to run from iPXE in some way.

You are indeed correct in that you can't send commandline options to the cmdline inside wimboot.
I have an idea for how this can be accomplished, it will get clunky/hacky

The best is probably to reach out to Intel about this.

**Comment by mcb30** [comment, marked as answer]
@bshogeman You could also try building iPXE with `DEBUG=efi_wrap`: this will show the boot services calls made by `iDmiEditEfi64.efi` and may give some indication as to the cause of failure.

**Reply by mcb30** [reply]
@bshogeman I have tested this myself in qemu.  The relevant final lines of output are:
```
OpenProtocol ( PciRoot(0x0)/Pci(0x3,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, 752f3136-4e16-4fdc-a22a-e5f46812f4ca, PciRoot(0x0)/Pci(0x3,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, <NULL>, G ) = UNSUPPORTED ( 0x0 ) -> 0x68426ad
OpenProtocol ( PciRoot(0x0)/Pci(0x3,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, 47c7b223-c42a-11d2-8e57-00a0c969723b, PciRoot(0x0)/Pci(0x3,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, <NULL>, G ) = UNSUPPORTED ( 0x0 ) -> 0x68426ea
```

These indicate that `iDmiEditEfi64.efi` is trying to open two EFI protocols on the image handle: `752f3136-4e16-4fdc-a22a-e5f46812f4ca` and `47c7b223-c42a-11d2-8e57-00a0c969723b`.  These are `EFI_SHELL_PARAMETERS_PROTOCOL_GUID` and `SHELL_INTERFACE_PROTOCOL_GUID`.  It therefore looks as though `iDmeEditEfi64.efi` is designed specifically to run from within the UEFI shell, and is not a generic UEFI application.

**Reply by NiKiZe** [reply]
Did not see mcb30s update before posting this, sorry for the noise.

```bash
iPXE> chain http://b800.org/idmi/iDmiEditEfi64.efi
http://b800.org/idmi/iDmiEditEfi64.efi... ok
WRAP PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi at base 0x7e3dd000 has protocols:
HANDLE PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi LoadedImageDevicePath supported
HANDLE PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi LoadedImage supported
HANDLE PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi LoadedImage opened 2x (G) by ipxe.efi for <NULL>
WRAP PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi parent ipxe.efi
WRAP PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi device PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)
WRAP PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi file iDmiEditEfi64.efi
OpenProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, 6302d008-7f9b-4f30-87ac-60c9fef5da4e, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, <NULL>, G ) = UNSUPPORTED ( 0x0 ) -> 0x7e40023a
LocateProtocol ( 6302d008-7f9b-4f30-87ac-60c9fef5da4e, 0x0 ) = NOT_FOUND ( 0x0 ) -> 0x7e40025c
OpenProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, 752f3136-4e16-4fdc-a22a-e5f46812f4ca, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, <NULL>, G ) = UNSUPPORTED ( 0x0 ) -> 0x7e40029f
OpenProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, 47c7b221-c42a-11d2-8e57-00a0c969723b, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, <NULL>, G ) = UNSUPPORTED ( 0x0 ) -> 0x7e40008b
LocateHandle ( ByProtocol, 47c7b221-c42a-11d2-8e57-00a0c969723b, 0x0, 0 ) = NOT_FOUND ( 0 ) -> 0x7e4000dd
OpenProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, 47c7b223-c42a-11d2-8e57-00a0c969723b, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, <NULL>, G ) = UNSUPPORTED ( 0x0 ) -> 0x7e400319
LocateProtocol ( 0fd96974-23aa-4cdc-b9cb-98d17750322a, 0x0 ) = 0 ( 0x7edbcb80 ) -> 0x7e3fe448
LocateProtocol ( ef9fc172-a1b2-4693-b327-6d32fc416042, 0x0 ) = 0 ( 0x7edbcba8 ) -> 0x7e3fe465
LocateProtocol ( 587e72d7-cc50-4f79-8209-ca291fc1a10f, 0x0 ) = 0 ( 0x7edbcc00 ) -> 0x7e3fe482
LocateProtocol ( HiiFont, 0x0 ) = 0 ( 0x7edbcb08 ) -> 0x7e3fe49f
LocateProtocol ( 31a6406a-6bdf-4e46-b2a2-ebaa89c40920, 0x0 ) = 0 ( 0x7edbcb28 ) -> 0x7e3fe26a
OpenProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, 752f3136-4e16-4fdc-a22a-e5f46812f4ca, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, <NULL>, G ) = UNSUPPORTED ( 0x0 ) -> 0x7e3fe6ad
OpenProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, 47c7b223-c42a-11d2-8e57-00a0c969723b, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/iDmiEditEfi64.efi, <NULL>, G ) = UNSUPPORTED ( 0x0 ) -> 0x7e3fe6ea
Could not boot: Error 0x7f0482ff (http://ipxe.org/7f0482ff)
iPXE>
```

When running thru shell from iPXE:
```
WRAP PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi file \iDmiEditEfi64.efi
OpenProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi, LoadedImage, Shell, <NULL>, G ) = 0 ( 0x7e5cf040 ) -> 0x7e3d890e
LocateDevicePath ( SimpleFileSystem, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi ) = 0 ( \iDmiEditEfi64.efi, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1) ) -> 0x7df1bdf1
LocateDevicePath ( SimpleFileSystem, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1) ) = 0 ( , PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1) ) -> 0x7df1beb0
InstallProtocolInterface ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi, 752f3136-4e16-4fdc-a22a-e5f46812f4ca, 0, 0x7fe96cb8 ) = 0 ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi ) -> 0x7df1e19e
StartImage ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi ) OpenProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi, 6302d008-7f9b-4f30-87ac-60c9fef5da4e, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi, <NULL>, G ) = UNSUPPORTED ( 0x0 ) -> 0x7e3d890e
LocateProtocol ( 6302d008-7f9b-4f30-87ac-60c9fef5da4e, 0x0 ) = 0 ( 0x7dfa35a0 ) -> 0x7de5a25c
OpenProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi, 752f3136-4e16-4fdc-a22a-e5f46812f4ca, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi, <NULL>, G ) = 0 ( 0x7fe96cb8 ) -> 0x7e3d890e
HandleProtocol ( ConIn, SimpleTextInputEx ) = 0 ( 0x7ec3ad00 ) -> 0x7df2c1ab0 ) -> 0x7de58448
Shell> WaitForEvent ( 0x7ecd2c98 ) -b327-6d32fc416042, 0x0 ) = 0 ( 0x7edbcba8 ) -> 0x7de58465
LocateProtocol ( 587e72d7-cc50-4f79-8209-ca291fc1a10f, 0x0 ) = 0 ( 0x7edbcc00 ) -> 0x7de58482
LocateProtocol ( HiiFont, 0x0 ) = 0 ( 0x7edbcb08 ) -> 0x7de5849f
LocateProtocol ( 31a6406a-6bdf-4e46-b2a2-ebaa89c40920, 0x0 ) = 0 ( 0x7edbcb28 ) -> 0x7de5826a
OpenProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi, 752f3136-4e16-4fdc-a22a-e5f46812f4ca, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi, <NULL>, G ) = 0 ( 0x7fe96cb8 ) -> 0x7e3d890e
LocateDevicePath ( SimpleFileSystem, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1) ) = 0 ( , PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1) ) -> 0x7df1dbc7
HandleProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1), DevicePath ) = 0 ( 0x7e2f7c28 ) -> 0x7df2d058
LocateDevicePath ( SimpleFileSystem, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi ) = 0 ( \iDmiEditEfi64.efi, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1) ) -> 0x7df1ca78
OpenProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1), SimpleFileSystem, Shell, <NULL>, G ) = 0 ( 0x7e2e3a70 ) -> 0x7e3d890e
OpenProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1), DevicePath, Shell, <NULL>, G ) = 0 ( 0x7e2f7c28 ) -> 0x7e3d890e
LocateDevicePath ( SimpleFileSystem, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi ) = 0 ( \iDmiEditEfi64.efi, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1) ) -> 0x7df1e4e6
HandleProtocol ( ConIn, SimpleTextInputEx ) = 0 ( 0x7ec3ad00 ) -> 0x7df2c1ab
This tool is not supported on this system.
HandleProtocol ( ConIn, SimpleTextInputEx ) = 0 ( 0x7ec3ad00 ) -> 0x7df2c1ab
49 - Error: A platform condition has prevented executing.
CloseProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi, 6302d008-7f9b-4f30-87ac-60c9fef5da4e, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi, <NULL> ) = NOT_FOUND -> 0x7de58574
CloseProtocol ( PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi, 752f3136-4e16-4fdc-a22a-e5f46812f4ca, PciRoot(0x0)/Pci(0x2,0x0)/MAC(525400123456,0x1)/\iDmiEditEfi64.efi, <NULL> ) = 0 -> 0x7de585a3
= 0x49 -> 0x7df1e304
UninstallProtocolInterface ( UNKNOWN<0x7e5d0d98,752f3136-4e16-4fdc-a22a-e5f46812f4ca>, 752f3136-4e16-4fdc-a22a-e5f46812f4ca, 0x7fe96cb8 ) = 0 -> 0x7df1e338
```

**Reply by mcb30** [reply]
You should therefore be able to run it via the shell using something like:
```
#!ipxe
imgfetch iDmiEditEfi64.efi
imgfetch startup.nsh
boot Shell.efi
```
where `startup.nsh` includes the command to invoke `iDmiEditEfi64.efi`.  This is convoluted, but is necessary since `iDmiEditEfi64.efi` is not a real UEFI application.

**Comment by bshogeman** [comment]
That method works!!  iDmiEditEfi64.efi  is loaded and executed.
I saw .nsh scripts support parameters like %1 %2 ..
But I must lookin if I can use iPXE to pass them trough.

Otherwise also not a problem.. Then it must be a PHP script on the server site to generate a startup.nsh file and on the iPXE script site sending the parameters with te request for the file like http://server/boot.php?a=${asset}&s=${serial}  That is doable.

A real big thanks for the solution and also making me see there is much more possible with nsh. I never used it before but it is almost a real old school batch file with great power.

I was already searching for a debug option in ipxe for efi but: DEBUG=efi  didn't work it had to be DEBUG=efi_wrap  I see in this forum.
But really thanks to you both for the solution. I would never found it myself. To focused on that it must be an simple efi file not knowing there can be different types.

The description/comment lines also helped a lot in understanding why. Tnx
